gadget/update.go

```diff
@@ -326,6 +327,95 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
 	return nil
 }
 
+// DiskVolumeDeviceTraitsForDevice takes a gadget volume and an expected
+// disk device path and builds up the traits for that device. It validates that
+// the specified volume can be laid out exactly to the disk device, returning an
+// error if the two are not compatible.
+func DiskVolumeDeviceTraitsForDevice(laidOutVolume *LaidOutVolume, dev string) (res DiskVolumeDeviceTraits, err error) {
```

Comment:
I think that the variable name `laidOutVolume` is quite confusing. The doc comment above seems to imply that this volume is not yet laid out ("it validates that the specified volume *can* be laid out"). Some naming alternatives:
- plannedLayout
- desiredLayout
- computedLayout

Reply:
Sorry if this is confusing, but I will respond to this and the other comment about EnsureLayoutCompatibility here together...

Well the laidOutVolume variable here really is not "laid out" inside this function and I suppose my doc-comment is not quite specific enough; really we are ensuring that the laidOutVolume is compatible with the onDiskVolume we construct for the device (that's what I referred to as "laying out", like fitting the pieces in a puzzle together). I think doing the validtion here and not as part of the caller makes sense because in all the callee sites, we will first start with a gadget.yaml, lay that out to get the abstract "LaidOutVolume", then we use some guesses to try and find a disk device which could match this LaidOutVolume, and then we will proceed to validate that it does indeed match the LaidOutVolume, and then finally return the DeviceTraits for it.

The issue with doing the validation elsewhere is a bit that if we didn't do the compatibility check here, then we would need to get the OnDiskVolume independently from this function, then run the compatibility check, then pass the OnDiskVolume to this function to build the traits and with this we would end up duplicating all the code which gets the OnDiskVolume for a device and the compatibility checking.

If we went with your proposal then we would end up with the following:
```go
func DiskVolumeDeviceTraitsForVolume(disk *OnDiskVolume) ... {
    // stuff
}
```

and have this style of code in a few places:

```go
dev, err := guessDiskDeviceForLayout(laidOutVolume)
if err != nil {
    return err
}

onDiskVol, err := OnDiskVolumeForDevice(dev)
if err != nil {
    return err
}

if err := EnsureLayoutCompatibility(laidOutVolume, onDiskVol); err != nil {
    return err
}

traits, err := DiskVolumeDeviceTraitsForVolume(onDiskVol)
if err != nil {
    return err
}
```

which to me feels more cumbersome. I also tend to like the fact that the helper is the one doing the validation check because it makes it harder to forget to do the validation when writing code because the helper is already doing it.

If others feel the same way about this function though, we can split the validation out.

However some alternative names which might be more clear than the existing one:

* `DiskTraitsIfCompatible(...)`
* `DiskTraitsWithValidation(...)`
* `DiskTraitsAndValidate(...)`
* `GetDiskTraitsAndValidate(...)`

perhaps @pedronis can comment on this

Reply:
...AndValidate is a pattern we use already in gadget. Anyway I agree that the function name and the comments currently are not accurate enough, with the name as committed I wouldn't expect this to do any sort of validation.

Reply:
I went with DiskTraitsFromDeviceAndValidate